compiler/src/dotty/tools/dotc/transform/SetRootTree.scala

```diff
@@ -0,0 +1,41 @@
+package dotty.tools.dotc.transform
+
+import dotty.tools.dotc.CompilationUnit
+import dotty.tools.dotc.ast.tpd
+import dotty.tools.dotc.core.Contexts.Context
+
+/** Set the `rootTreeOrProvider` property of class symbols. */
+class SetRootTree extends MacroTransform {
+
+  override val phaseName: String = SetRootTree.name
+  override def isRunnable(implicit ctx: Context) =
+    super.isRunnable && ctx.settings.YretainTrees.value
+
+  override protected def newTransformer(implicit ctx: Context): Transformer = {
+    new Transformer {
+      override def transform(tree: tpd.Tree)(implicit ctx: Context): tpd.Tree = {
+        tree match {
+          case pkg: tpd.PackageDef =>
+            super.transform(pkg)
+          case td: tpd.TypeDef =>
+            if (td.symbol.isClass) {
+              val sym = td.symbol.asClass
+              tpd.sliceTopLevel(ctx.compilationUnit.tpdTree, sym) match {
+                case (pkg: tpd.PackageDef) :: Nil =>
+                  sym.rootTreeOrProvider = pkg
```

Comment:
The `rootTreeOrProvider` should only be set in the top level class

Reply:
I think it's the case already since this no longer traverses the body of classes, no?

Reply:
I misinterpreted the code. It is fine.